frame/balances/src/lib.rs

```diff
@@ -651,15 +652,20 @@ impl<T: Config<I>, I: Instance> Module<T, I> {
 		who: &T::AccountId,
 		f: impl FnOnce(&mut AccountData<T::Balance>, bool) -> Result<R, E>
 	) -> Result<R, E> {
-		T::AccountStore::try_mutate_exists(who, |maybe_account| {
+		let result = T::AccountStore::try_mutate_exists(who, |maybe_account| {
 			let is_new = maybe_account.is_none();
 			let mut account = maybe_account.take().unwrap_or_default();
 			f(&mut account, is_new).map(move |result| {
 				let maybe_endowed = if is_new { Some(account.free) } else { None };
-				*maybe_account = Self::post_mutation(who, account);
-				(maybe_endowed, result)
+				let maybe_account_maybe_dust = Self::post_mutation(who, account);
+				*maybe_account = maybe_account_maybe_dust.0;
+				(maybe_endowed, maybe_account_maybe_dust.1, result)
 			})
-		}).map(|(maybe_endowed, result)| {
+		});
+		result.map(|(maybe_endowed, maybe_dust, result)| {
+			if let Some(dust) = maybe_dust {
+				T::DustRemoval::on_unbalanced(dust);
```

Comment:
in `fn transfer` and `fn repatriate_reserved`, `try_mutate_account` is called inside another `try_mutate_account`.

Thus `on_unbalanced` cannot change the balance of the `dest` of a transfer. and cannot change the balance of the `beneficiary` of a repatriate_reserved.

Maybe we need another PR, or change transfer and repatriate_reserved here to fix the issue completely

Reply:
That's a separate issue though to do with nested `mutate_account`s, whereas this fixes the issue of a single `mutate_account` where the `on_unbalanced` happens to touch the account being mutated.

I fear with nested `mutate_account`s, we'll always see issues like this and we'll probably need to alter the pattern to fix it properly. Something like:

```
{
  let u32_wrapper_value: SomeSpecialType<u32> = U32StorageItem::get_mutable();
  *u32_wrapper_value += 1;
  // calling `.flush()` would flush changes into the storage item
  u32_wrapper_value.flush();
  // other stuff could be written in afterwards
  *u32_wrapper_value = 0;
}
// After the `SomeSpecialType` is dropped, any pending changes are written, like with `.flush()`
```

Reply:
yes I agree, maybe nested `mutate_account` calls of pallet-balances can just be rewritten, like with https://github.com/paritytech/substrate/pull/7783
Or introducing another abstraction